### ISSUE TITLE
Paragraph: Color combinations can fail contrast requirements

### ISSUE THREAD

**Issue by nic-bertino** [WordPress/gutenberg#2381]
## Issue Overview
Certain color combinations with the Cover Text block fail to meet contrast ratios for [WCAG 2.0/AA](https://www.w3.org/TR/WCAG/#visual-audio-contrast).

## Current Behavior
1. Select the same color for `color` and `background-color` to create a 1.0 contrast ratio.
2. Most other combinations fail, such as [this](http://leaverou.github.io/contrast-ratio/#%20rgb%286%2C%20147%2C%20227%29-on-rgb%28235%2C%2020%2C%2076%29) will be difficult to read for sighted users as well.

## Possible Solution
1. Don't allow a 1.0 contrast by limiting the selection of colors that are the same
2. Use a tool to determine contrast ratios and promote color combinations that meet WCAG 2.0/AA requirements.

Edit: Cover text's UI controls were merged into the Paragraph controls in 1.0. The title has been updated accordingly.

**Comment by 00travelgirl00** [comment]
Just wanted to open the same ticket. 

Not sure if this is possibile, but I would love to have an indicator for the contrast ratio right in the Gutenberg for editing options like this. 

Because this would also teach and make people aware of problems like this.

**Comment by joedolson** [comment]
I'd like to propose a feature for the color picker: color contrast suggestions. If the color picker could pre-select colors that will meet color contrast guidelines, this would go a long ways to ensuring best practices for accessibility in content. 

The actual functionality might vary depending on context, but as an accessibility tool it would be invaluable.

**Comment by nic-bertino** [comment]
Just a note - in 1.0, paragraphs now have color pickers, which have the same UI control:

<img width="1205" alt="screenshot 2017-08-29 11 49 32" src="https://user-images.githubusercontent.com/883093/29838481-25ef2c16-8cb0-11e7-9305-539143fa6581.png">

**Comment by BinaryMoon** [comment]
The colours chosen have a lack of contrast. The colour white is also absent - and since themes can have dark backgrounds as well as light - I think it is an essential option.

The default colours chosen are nice and bold but as mentioned the combinations you can make that are readable are few.

If you greyscale the palette you can see the problem.

![lack-of-contrast](https://user-images.githubusercontent.com/1004130/29840427-e36f5f42-8cf9-11e7-8bb7-a3e2e5c71409.jpg)

There's 2 dark colours (1 of which is black) and 1 light colour. The rest have a middling luminance so any combination that uses 2 of those colours will be almost unreadable.

I also don't think the effect used for displaying the selected colour is very good. You can barely see the chosen colour. I'd use the second border effect - as seen in the colour focus - to signify the chosen colour.

**Comment by afercia** [comment]
@BinaryMoon I see your point. Slightly unrelated though, seems to me it's more related to the general functionality. Would you mind opening a new issue? Just to keep things separated. Also please see https://github.com/WordPress/gutenberg/issues/2381 and consider, as far as I know, themes should be able to extend the list of colors.

**Comment by 00travelgirl00** [comment]
I did some mockup for how it could look like. I prefer the version in the sidebar. Maybe with an option to explain the color contrast a bit better. 

Also the content stays clean in this way and all the settings are in the sidebar. 

What do you think? 

![dont_pass_button](https://user-images.githubusercontent.com/21318377/30292436-72656746-9736-11e7-9edd-7313383e2c6d.jpg)
![pass_button](https://user-images.githubusercontent.com/21318377/30292440-74741064-9736-11e7-9013-49dd6b3a60c1.jpg)
![dont_pass_sidebar](https://user-images.githubusercontent.com/21318377/30292441-762ab782-9736-11e7-9484-e08a5de3c082.jpg)
![pass_sidebar](https://user-images.githubusercontent.com/21318377/30292443-781a0cbe-9736-11e7-8b40-adb2c03f3fd4.jpg)

**Comment by nic-bertino** [comment]
Rebounding off of @00travelgirl00's mock, here's what I've quickly come up with:

![color combinations](https://user-images.githubusercontent.com/883093/30337972-e0156742-979e-11e7-9077-1e33e1a284af.png)

**Comment by BinaryMoon** [comment]
I like the idea of selecting a background colour, and then having WordPress automatically calculate the contrasting text colour. I always think white and black on coloured backgrounds look best anyway.

You'd then end up with something like @nic-bertino's suggestion above. You can then have all the bold colours you like, safe in the knowledge the text will be readable.

**Comment by fisicx** [comment]
Suggestions are good but if you want to be really useful you need something like this: http://paletton.com

And I may have found a bug. I wanted to change the colour of two words in a block. I highlit the two and used the colour selector but this changed the whole block. The doesn't appear to be a way to set the colour of anything less than a whole block

**Comment by afercia** [comment]
As the previous comment highlights, I think Gutenberg is going to put in the users hand a tool that will be abused in several ways. Plenty of colors everywhere. At this point, why not add an option to make all text Comic Sans? 🙂 
Jokes apart, I'm increasing the priority of this issue to high, since it is an accessibility priority.

At the very least, Gutenberg should warn users when a chosen color combination doesn't meet a sufficient contrast ratio, in the same way it warns when the headings hierarchy is not correct (see the Table of Contents).

See also the related discussion during latest accessibility meeting on Slack: https://wordpress.slack.com/archives/C02RP4X03/p1506362229000557

**Comment by nic-bertino** [comment]
See also #2770, which suggests iterations on the color picker UI pattern. I feel like we're working in parallel.

**Comment by karmatosed** [comment]
For now, removing priority high as that should be only for escalated issues not for enhancements. Not having a high priority doesn't mean it won't get done, but we should be cautious of assigning it.

**Comment by afercia** [comment]
@karmatosed we've been asked from you to mark the accessibility issues we consider a priority. This is one of them, and for this reason was marked with both the accessibility and high priority labels. There are also other issues with both labels, to indicate "a11y priority".

Please let us know which labels we should use to mark the "must have" accessibility improvements and fixes, as this should have an indication of higher priority. Thanks.

**Comment by karmatosed** [comment]
I totally understand marking priorities but I did not ask you to mark as high, so lets keep with the accessibility label we have now. We will tackle must haves through communication, no need for labels at this point. I asked you to let me know a11y priorities and that totally is how we should start doing this. Feel free to bring them to the weekly meeting also, there is space.

**Comment by afercia** [comment]
Why not to use a new label Something like `a11y-must-have` or similar. I'd like to avoid to maintain separate lists...

**Comment by afercia** [comment]
Here's the current list of Accessibility + Priority HIgh:
https://github.com/WordPress/gutenberg/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20label%3AAccessibility%20%20label%3A%22Priority%20High%22%20

**Comment by karmatosed** [comment]
My thinking (open to other's insights), is that yet another label is yet another thing to get lost in. The milestone I have labelled is one both leads are taking note of, which means it's not being lost.

@mtias what is your feeling here? I think a lot of those aren't 'high priority' for the project - which is the issue with you labelling everything high priority right? Let's take a step back. We have that list now and can remove the priority label and get into a milestone.

**Comment by samikeijonen** [comment]
> At the very least, Gutenberg should warn users when a chosen color combination doesn't meet a sufficient contrast ratio, in the same way it warns when the headings hierarchy is not correct (see the Table of Contents).

+100 for this. @joedolson probably knows [is this the correct formula for calculating the contrast](https://www.w3.org/TR/WCAG20-TECHS/G17.html).

**Comment by joedolson** [comment]
Yes; that's the correct formula.

**Comment by karmatosed** [comment]
Here's a suggestion of what this could look like. I think it should kick in when you select the 'second' color. If that's text first the message can move up. We can iterate on the language, but I think a 'warning' is right not an error and not a block.

![accessibilitycheck-colors](https://user-images.githubusercontent.com/253067/31401223-2f601b7e-adea-11e7-8cfc-330960699cef.png)

**Comment by melchoyce** [comment]
When we iterate on copy, let's be sure to offer up some tips for improvement, like:

> This color combination will be hard for people to read. Try using a [darker/lighter] background color.

**Comment by BinaryMoon** [comment]
Is there a possibility to remove the foreground colour choice entirely? If WordPress were to select the foreground colour (either white or black) based upon the background colour then accessibility would be ensured.

This would also align with the Decisions not options philosophy.

**Comment by joedolson** [comment]
In principle, that would be effective, and I've done it myself in themes. However, it does create its own set of problem,largely because white and black are not necessarily the best choices for contrast, since max contrast is frequently not the best choice for accessibility.

It does simplify the UI to eliminate the choice of text color, but there are many, many totally valid combinations of two colors that meet accessibility guidelines, and I don't feel that it's a good decision by the project to determine by fiat that those options will not be supported.

**Comment by BinaryMoon** [comment]
@joedolson - no problem. Thought it was worth suggesting since I think it's best to have the minimum options possible.
